### ISSUE TITLE
if() and eval() produce incorrect results in RETURN section of MATCH

### ISSUE THREAD

**Issue by obi1kenobi** [orientechnologies/orientdb#6606]
## OrientDB Version, operating system, or hardware.
- v2.0 SNAPSHOT[ ] - .18[ ] .17[ ] .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.1 SNAPSHOT[ ] - .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.2 SNAPSHOT[ ] - .7[x] .rc1[ ] .beta2[ ] .beta1[ ]

## Operating System
- [x] Linux
- [x] MacOSX
- [ ] Windows
- [ ] Other Unix
- [ ] Other, name? 

## Expected behavior and actual behavior
Expected:
- The `RETURN` section of `MATCH` should be equivalent in functionality to the section between the `SELECT` and `FROM` section of a `SELECT` statement. For example, using `if()` and `eval()` is allowed and produces equivalent results in both of these.
- Regardless of what is in the `RETURN` section of a `MATCH`, the number of rows returned should not change.

Actual:
Using `if()` and `eval()` in the `RETURN` section of a `MATCH` does not appear to work correctly, and changes the number of rows relative to when the query ends in `RETURN $matches`. See repro below for concrete examples.

## Steps to reproduce the problem

Setup:
```
CREATE CLASS V1 EXTENDS V
CREATE PROPERTY V1.name String
CREATE VERTEX V1 SET name = 'foo'
CREATE VERTEX V1 SET name = 'bar'
```

First query, without `if()` or `eval()`, performs as expected:
```
MATCH {
    class: V1,
    as: p
}
RETURN $matches
```
This returns two rows, and the `p` column has the correct values `#12:0` and `#12:1`.

Use `if()` and `eval()` with constants in the "true-value" and "false-value" of the `if()`:
```
MATCH {
    class: V1,
    as: p
}
RETURN if(eval("p.name == 'foo'"), 1, 2) AS b
```
This returns *a single row* rather than two rows as the previous query, and the `b` column has the value `2` indicating a `false` condition. However, one of the vertices should have satisfied the condition. If I had to guess, my money would be on the alias `p` not being defined inside the `eval()`.

Use `if()` and `eval()` with alias-based expressions in the "true-value" and "false-value" of the `if()`:
```
MATCH {
    class: V1,
    as: p
}
RETURN if(eval("p.name == 'foo'"), p.name, p.name) as b
```
This once again returns a single row rather than two rows, and the `b` column contains the null value, even though both the "true-value" and the "false-value" were the `name` of the `V1` vertex. It again appears as if the alias `p` is not defined inside the scope of the `if()`.

Rewriting these same queries as:
```
SELECT <selections_with_ifs_and_evals_here> FROM (
    MATCH {
        class: V1,
        as: p
    } RETURN $matches
)
```
always produces the expected, correct results -- two rows with the proper "true-value" and "false-value" in each row.

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

I just pushed a fix for this, so now the first case works fine.
Anyway, please consider that the standard behavior of MATCH statement is to return distinct patterns removing duplicates, so the `if(eval("p.name == 'foo'"), p.name, p.name) as b` will still return only one row. To have all the results (not distinct) you have to use an outer SELECT

Thanks

Luigi

**Comment by obi1kenobi** [comment]
Hi @luigidellaquila, thanks for the fix!

I'm not sure I understand why the `if(eval("p.name == 'foo'"), p.name, p.name) as b` would return only one row. Two vertices with different names (`'foo'` and `'bar'`) should be produced by the MATCH statement, and since the output of the `if(eval())` is always `p.name`, shouldn't the result be two rows with `'foo'` and `'bar'`?

In other words, the problem with the second example was that the returned value was `null` rather than `'foo'` or `'bar'`, and there was no vertex with `null` name. I understand that MATCH does deduplication of the returned results, but there were no duplicates here.

**Comment by obi1kenobi** [comment]
Ah I see the confusion now -- in your commit, the second test case uses a literal constant `'foo'` in both the if-true and the if-false values. However, note that this is not what my test case has -- it has `p.name` in each, which changes per-row since there are two vertices with distinct names that `p` applies to.

Would you be able to verify that your fix also fixes the `null` return problem my second testcase pointed out?

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

Sorry for the confusion, my bad, you are absolutely right.
If two records have different names, the result will definitely be two rows, not one.

Thanks

Luigi

**Comment by obi1kenobi** [comment]
No problem :) As the cause of the erroneous `null` return when `p.name` is never `null` may be different than the one causing the `if(eval())` problem, would you mind double-checking that my query indeed works with your patch?

My project unfortunately requires both of these issues to be resolved before I can use MATCH, and I'd like to make sure I can switch in 2.2.11 as the performance improvement is huge!

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

Ok, reopening for a double check, I'll give you a feedback in short

Thanks

Luigi

**Comment by obi1kenobi** [comment]
Thanks, much appreciated!

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

I just double-checked, null values are correctly handled (like all the other values)

Closing, thanks!

Luigi

**Comment by obi1kenobi** [comment]
Hi @luigidellaquila, I'm afraid there might have been another misunderstanding. The problem I reported isn't with the *handling* of null values, it's the fact that null values were being returned *instead of the correct, non-null value* if a MATCH selection alias was used in the return values of the `if()` operator. To be clear, `p.name` *was not null* but the query returned a result *claiming* that `p.name` was null.

`if(eval("p.name == 'foo'"), p.name, p.name) as b` should always return `p.name` regardless of what the name actually is. In my testcase, `p.name` *was not null* and yet the returned value for `b` was null. Could you please verify that your fix indeed fixes the above?

**Comment by luigidellaquila** [comment]
Hi @obi1kenobi 

Ah, ok, got it. The main fix actually was on that point, the `p` was not correctly evaluated, so it returned null for both the eval expression and the p.name in the `if` expression.

I just fixed the problem passing correctly the context to the functions, so now everything is correctly evaluated.

In short, that is exactly what I fixed, all the rest was a consequence ;-)

Thanks

Luigi
